UefiPayloadPkg/Library/LbParseLib/LbParseLib.c

```diff
@@ -0,0 +1,182 @@
+#include <IndustryStandard/SmBios.h>
+#include <Library/BaseLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/BlParseLib.h>
+#include <Library/DebugLib.h>
+#include <Library/IoLib.h>
+#include <Library/PcdLib.h>
+#include <Linuxboot.h>
+#include <Uefi/UefiBaseType.h>
+
+// Retrieve UefiPayloadConfig from Linuxboot's uefiboot
+UefiPayloadConfig* GetUefiPayLoadConfig() {
+  UefiPayloadConfig *Config =
+      (UefiPayloadConfig*)(UINTN)(PcdGet32(PcdPayloadFdMemBase) - SIZE_64KB);
+  if (Config->Version != UEFI_PAYLOAD_CONFIG_VERSION) {
+    DEBUG((DEBUG_ERROR, "Expect payload Config version: %d, but get %d\n",
+           UEFI_PAYLOAD_CONFIG_VERSION, Config->Version));
+    CpuDeadLoop ();
+  }
+  return Config;
+}
+
+// Align the address and add memory rang to MemInfoCallback
+void AddMemoryRange(
+    IN BL_MEM_INFO_CALLBACK MemInfoCallback,
+    IN UINTN start,
+    IN UINTN end,
+    IN int type) {
+  MEMROY_MAP_ENTRY MemoryMap;
+  UINTN AlignedStart;
+  UINTN AlignedEnd;
+  AlignedStart = ALIGN_VALUE(start, SIZE_4KB);
+  AlignedEnd = ALIGN_VALUE(end, SIZE_4KB);
+  // Conservative adjustment on Memory map. This should happen when booting from
+  // non UEFI bios and it may report a memory region less than 4KB.
+  if (AlignedStart > start && type != LINUXBOOT_MEM_RAM) {
+    AlignedStart -= SIZE_4KB;
+  }
+  if (AlignedEnd > end + 1 && type == LINUXBOOT_MEM_RAM) {
+    AlignedEnd -= SIZE_4KB;
+  }
+  MemoryMap.Base = AlignedStart;
+  MemoryMap.Size = AlignedEnd - AlignedStart;
+  MemoryMap.Type = type;
+  MemoryMap.Flag = 0;
+  MemInfoCallback(&MemoryMap, NULL);
+}
+
+/**
+  Acquire the memory information from the linuxboot table in memory.
+
+  @param  MemInfoCallback     The callback routine
+  @param  Params              Pointer to the callback routine parameter
+
+  @retval RETURN_SUCCESS     Successfully find out the memory information.
+  @retval RETURN_NOT_FOUND   Failed to find the memory information.
+
+**/
+RETURN_STATUS
+EFIAPI
+ParseMemoryInfo(
+    IN BL_MEM_INFO_CALLBACK MemInfoCallback,
+    IN VOID* Params
+    ) {
+  UefiPayloadConfig *Config;
+  int Index;
+
+  Config = GetUefiPayLoadConfig();
+
+  DEBUG((DEBUG_INFO, "MemoryMap #entries: %d\n", Config->NumMemoryMapEntries));
+
+  MemoryMapEntry* entry = &Config->MemoryMapEntries[0];
+  for (Index = 0; Index < Config->NumMemoryMapEntries; Index++) {
+    DEBUG((DEBUG_INFO, "Start: 0x%lx End: 0x%lx Type:%d\n", entry->Start,
+           entry->End, entry->Type));
+    AddMemoryRange(MemInfoCallback, entry->Start, entry->End, entry->Type);
+    entry++;
+  }
+  return RETURN_SUCCESS;
+}
+
+/**
+  Acquire acpi table and smbios table from linuxboot
+
+  @param  SystemTableInfo          Pointer to the system table info
+
+  @retval RETURN_SUCCESS            Successfully find out the tables.
+  @retval RETURN_NOT_FOUND          Failed to find the tables.
+
+**/
+RETURN_STATUS
+EFIAPI
+ParseSystemTable(
+    OUT SYSTEM_TABLE_INFO* SystemTableInfo
+    ) {

```

Comment:
please keep 2 spaces at the beginning.
  OUT SYSTEM_TABLE_INFO* SystemTableInfo
  ) {